w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,259 @@
+               continue
+
+            #this hack is to make w3af send request with unencoded qs
+            u =  str(mu.get_uri())
+            ptest_url = URL(u.replace(REFLECT_TOKEN,
+                                      value +'&'+param+'='+REFLECT_TOKEN))
+            ptest_resp = self._uri_opener.GET(ptest_url)
+            ptest_resp_code = ptest_resp.get_code()
+            
+            if not ptest_resp_code == 200:
+                om.out.debug(u"Precedence check request to the uri %s "
+                             u"failed! "
+                             u"Got response code %s in response id %s\n"
+                             u"Unable to continue HPP detection. "
+                             u"Resource might be protected by a WAF,"
+                             % (ptest_url, ptest_resp_code, ptest_resp.id))
+                #no need to seek reflection if it is not OK
+                continue
+
+            ptest_body = self.strip_trash(ptest_resp.body)
+
+            try:
+                ptest_dp = parser_cache.dpc.get_document_parser_for(ptest_resp)
+            except BaseFrameworkException:
+                return
+
+            if fuzzy_equal(self.strip_trash(orig_response.body),
+                           ptest_body, EQUALITY_THRESHOLD):
+                #confirm vuln
+                if self.find_reflection( ptest_dp, value) == rp:
+                    self.precedence =  FIRST
+                    continue
+            if fuzzy_equal(body, ptest_body, EQUALITY_THRESHOLD):
+                #confirm vuln
+                if self.find_reflection( ptest_dp, REFLECT_TOKEN) == rp:
+                    self.precedence = LAST
+                    continue
+                #  report vulnerability
+            try:
+                reflection, link = self.find_reflection( ptest_dp, value, REFLECT_TOKEN)
+            except TypeError:
+                self.precedence == UNK
+
+            if reflection == rp:
+                self.precedence = CONCAT
+                desc = u'Concatenation of values of the two input parameters ' \
+                       u'with the same name \"%s\" detected inside HTML ' \
+                       u'code of response body.\nThe request url was %s.\n' \
+                       u'Got concatenated output reflected in this link %s\n' \
+                       u'This \"feature\" may be used in different WAF ' \
+                       u'bypassing techniques.'
+                desc = desc % (param, ptest_url, link)
+                print desc
+
+                i = Info(u'Input concatenation detected', desc, ptest_resp.id,
+                         self.get_name())
+                i.set_url(ptest_url)
+                self.kb_append(self, 'hpp', i)
+                return
+
+
+
+        if self.precedence == UNK:
+            om.out.debug(u'Unable to determine precedence for HPP detection.'
+                         u'Presumably not vulnerable to HPP.')
+
+            return
```

Comment:
Maybe this return should be a continue? Maybe the precedence was not possible to "calculate" with this QS parameter but will be with the next? Not sure about this... just asking.